File: js/quick_sort_1.js
/**
 * This implementation is based on the first example
 * on wikipedia https://en.wikipedia.org/wiki/Quicksort
 * @param {number[]} A
 * @param {number} lo
 * @param {number} hi
 */
function quicksort(A, lo, hi) {
    if (lo < hi) {
        const p = partition(A, lo, hi);
        quicksort(A, lo, p - 1)
        quicksort(A, p + 1, hi)
    }
}

/**
 * @param A
 * @param {number} lo
 * @param {number} hi
 * @returns {*}
 */
function partition(A, lo, hi) {
    let pivot = A[hi];
    let i = lo;
    // move between lo & hi
    for (let j = lo; j <= hi; j += 1) {
        if (A[j] < pivot) {
            swap(A, i, j);
            i += 1;
        }
    }
    swap(A, i, hi);
    return i;
}

function swap(A, i, j) {
    if (i === j) return; // do nothing if the indexes are equal
    const first = A[i];
    A[i] = A[j];
    A[j] = first;
}

const assert = require("assert");
const items  = [10, 9, 5, 3, 20, 18];
const sorted = [3, 5, 9, 10, 18, 20];
quicksort(items, 0, items.length-1);
assert.deepStrictEqual(items, sorted);

File: src/quick_sort_1.rs
///
/// This implementation is based on the first example
/// on wikipedia https://en.wikipedia.org/wiki/Quicksort
///
fn quicksort(a: &mut [usize], lo: usize, hi: usize) {
    if lo < hi {
        let pivot = partition(a, lo, hi);

        // this check is to prevent overflows on negative numbers
        if pivot > 0 {
            quicksort(a, lo, pivot - 1);
        }
        quicksort(a, pivot + 1, hi);
    }
}

fn partition(a: &mut [usize], lo: usize, hi: usize) -> usize {
    let pivot = a[hi];
    let mut i = lo;
    for j in lo..=hi {
        if a[j] < pivot {
            swap(a, i, j);
            i += 1;
        }
    }
    swap(a, i, hi);
    i
}

fn swap(a: &mut [usize], lo: usize, hi: usize) {
    a.swap(lo, hi);
}

#[test]
fn test_quick_sort_1() {
    let mut items: Vec<usize> = vec![10, 9, 5, 3, 20, 18];
    let len = items.len();
    quicksort(&mut items, 0, len - 1);
    assert_eq!(items, &[3, 5, 9, 10, 18, 20])
}
